lib/unison-sqlite/src/Unison/Sqlite/Transaction.hs

```diff
@@ -67,13 +71,33 @@ newtype Transaction a
 
 -- | Run a transaction on the given connection.
 runTransaction :: MonadIO m => Connection -> Transaction a -> m a
-runTransaction conn@(Connection _ _ conn0) (Transaction f) =
-  -- TODO some sensible retry logic
-  liftIO do
-    Sqlite.execute_ conn0 "BEGIN"
-    result <- f conn
-    Sqlite.execute_ conn0 "COMMIT"
+runTransaction conn (Transaction f) = liftIO do
+  uninterruptibleMask \restore -> do
+    Connection.execute_ conn "BEGIN"
+    result <-
+      try (restore (f conn)) >>= \case
+        Left exception -> do
+          rollback
```

Comment:
What happens if this throws an exception while we're inside a mask?
Legitimately asking here, I'm not sure what happens.

Is there any chance this blocks? We need to be really careful here.

Reply:
> What happens if this throws an exception while we're inside a mask?

If it throws an exception, then it throws an exception - the masking state doesn't affect that.

> Is there any chance this blocks? We need to be really careful here.

Yes, not for long. It's not a blocking operation like `select` or anything. Perhaps if the file's write buffer needs flushing or something it would block momentarily.

Reply:
Looks like we may want to ignore errors on this rollback here since if SqliteBusy is returned it's possible the transaction has already been rolled back and this rollback would trigger an error.

```suggestion
          ignoringExceptions rollback
```

https://www.sqlite.org/lang_transaction.html

> If certain kinds of errors occur within a transaction, the transaction may or may not be rolled back automatically. 
> ...
> It is recommended that applications respond to the errors listed above by explicitly issuing a ROLLBACK command. If the transaction has already been rolled back automatically by the error response, then the ROLLBACK command will fail with an error, but no harm is caused by this.

Reply:
Good catch, thanks